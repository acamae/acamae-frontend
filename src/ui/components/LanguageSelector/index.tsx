import React, { useEffect, useRef } from 'react';
import { Button, Form } from 'react-bootstrap';
import { useTranslation } from 'react-i18next';

enum SupportedLngs {
  'es-ES' = 'es-ES',
  'en-GB' = 'en-GB',
}

const LanguageSelector: React.FC = () => {
  const { t, i18n } = useTranslation();
  const supportedLngs = Array.isArray(i18n.options.supportedLngs)
    ? i18n.options.supportedLngs.filter(lng => lng !== 'cimode')
    : ['es-ES', 'en-GB'];
  const currentLang = supportedLngs.includes(i18n.language) ? i18n.language : supportedLngs[0];
  const liveRegionRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    document.documentElement.lang = currentLang;
    if (liveRegionRef.current) {
      liveRegionRef.current.textContent = t('language.changed', {
        lng: t(`language.${currentLang}`),
      });
    }
  }, [currentLang, t]);

  const handleChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const newLang = event.target.value;
    if (newLang !== i18n.language && supportedLngs.includes(newLang)) {
      i18n
        .changeLanguage(newLang)
        .then(() => {
          localStorage.setItem('i18nextLng', newLang);
        })
        .catch(() => {
          if (liveRegionRef.current) {
            liveRegionRef.current.textContent = t('language.change_error');
          }
        });
    }
  };

  return (
    <div className="d-flex align-items-center language-picker">
      <Form className="language-picker__form" data-testid="language-selector">
        <Form.Group className="language-picker__group">
          <Form.Label
            htmlFor="language-picker-select"
            className="visually-hidden language-picker__label"
            data-testid="lang-select-label">
            {t('language.selector_label')}
          </Form.Label>
          <Form.Select
            id="language-picker-select"
            name="language-picker-select"
            value={currentLang}
            className="language-picker__select"
            aria-label={t('language.selector_label')}
            onChange={handleChange}
            data-testid="language-picker-select">
            {supportedLngs.map((lng: SupportedLngs) => (
              <option data-testid={`language-picker__option_${lng}`} key={lng} value={lng}>
                {t(`language.${lng}`)}
              </option>
            ))}
          </Form.Select>
          <span id="language-selector-desc" className="visually-hidden">
            {t('language.selector_hint')}
          </span>
          <div
            ref={liveRegionRef}
            aria-live="polite"
            aria-atomic="true"
            className="visually-hidden"
          />
          <Button
            type="submit"
            className="visually-hidden language-picker__button"
            aria-label="English, Select your language"
            aria-expanded="false"
            aria-controls="language-picker-dropdown">
            <span
              aria-hidden="true"
              className="language-picker__flag language-picker__flag--english"></span>
          </Button>
        </Form.Group>
      </Form>
    </div>
  );
};

export default LanguageSelector;
