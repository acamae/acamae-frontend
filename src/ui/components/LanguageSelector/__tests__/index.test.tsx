jest.mock('react-i18next');

import { screen, fireEvent, render, waitFor } from '@testing-library/react';
import { useTranslation } from 'react-i18next';

import LanguageSelector from '@ui/components/LanguageSelector';

let setItemSpy: jest.SpyInstance;
function setupUseTranslation({
  lang = 'es-ES',
  t = (str: string) => str,
  changeLanguage = createChangeLanguageMock(),
  supportedLngs = ['es-ES', 'en-GB'],
} = {}) {
  (useTranslation as jest.Mock).mockReturnValue({
    t,
    i18n: {
      language: lang,
      changeLanguage,
      options: { supportedLngs },
    },
  });
}

function createChangeLanguageMock(shouldResolve = true) {
  return jest.fn(() => (shouldResolve ? Promise.resolve() : Promise.reject(new Error('error'))));
}

function renderLanguageSelector() {
  return render(<LanguageSelector />);
}

describe('LanguageSelector', () => {
  beforeEach(() => {
    setItemSpy = jest.spyOn(Storage.prototype, 'setItem').mockImplementation(jest.fn());
  });

  afterEach(() => {
    setItemSpy.mockRestore();
  });

  it('Should render the selector with the current language selected (es)', () => {
    setupUseTranslation();
    renderLanguageSelector();
    const selectElement = screen.getByTestId('language-picker-select') as HTMLSelectElement;
    expect(selectElement).toBeInTheDocument();
    expect(selectElement.value).toBe('es-ES');
  });

  it('Should render the selector with the current language selected (en)', () => {
    setupUseTranslation({ lang: 'en-GB' });
    renderLanguageSelector();
    const selectElement = screen.getByTestId('language-picker-select') as HTMLSelectElement;
    expect(selectElement.value).toBe('en-GB');
  });

  it('Should call the changeLanguage function of the hook and localStorage.setItem when changing the language', async () => {
    const changeLanguageMock = createChangeLanguageMock(); // success
    setupUseTranslation({ changeLanguage: changeLanguageMock });
    renderLanguageSelector();
    const selectElement = screen.getByTestId('language-picker-select') as HTMLSelectElement;
    fireEvent.change(selectElement, { target: { value: 'en-GB' } });

    await waitFor(() => {
      expect(changeLanguageMock).toHaveBeenCalledWith('en-GB');
      expect(setItemSpy).toHaveBeenCalledWith('i18nextLng', 'en-GB');
    });
  });

  it('Should not call changeLanguage or localStorage.setItem if the same language is selected', async () => {
    const changeLanguageMock = createChangeLanguageMock(false); // error
    setupUseTranslation({ changeLanguage: changeLanguageMock });
    renderLanguageSelector();
    const selectElement = screen.getByTestId('language-picker-select') as HTMLSelectElement;
    fireEvent.change(selectElement, { target: { value: 'es-ES' } });

    await waitFor(() => {
      expect(changeLanguageMock).not.toHaveBeenCalled();
      expect(setItemSpy).not.toHaveBeenCalled();
    });
  });

  it('Should have a hidden label for accessibility', () => {
    setupUseTranslation();
    renderLanguageSelector();
    const label = screen.getByLabelText('language.selector_label');
    expect(label).toBeInTheDocument();
  });

  it('Should not call changeLanguage with unsupported language', async () => {
    const changeLanguageMock = createChangeLanguageMock(false);
    setupUseTranslation({ changeLanguage: changeLanguageMock });
    renderLanguageSelector();
    const selectElement = screen.getByTestId('language-picker-select') as HTMLSelectElement;
    fireEvent.change(selectElement, { target: { value: 'fr-FR' } });

    await waitFor(() => {
      expect(changeLanguageMock).not.toHaveBeenCalled();
      expect(selectElement.value).toBe('es-ES');
    });
  });

  it('Should show the available options', () => {
    setupUseTranslation();
    renderLanguageSelector();
    expect(screen.getByTestId('language-picker__option_es-ES')).toBeInTheDocument();
    expect(screen.getByTestId('language-picker__option_en-GB')).toBeInTheDocument();
  });

  it('snapshot: render of the selector', () => {
    setupUseTranslation();
    renderLanguageSelector();
    const { asFragment } = renderLanguageSelector();
    expect(asFragment()).toMatchSnapshot();
  });
});
