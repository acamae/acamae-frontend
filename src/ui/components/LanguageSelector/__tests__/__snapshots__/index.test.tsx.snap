// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageSelector snapshot: render of the selector 1`] = `
<DocumentFragment>
  <div
    class="d-flex align-items-center language-picker"
  >
    <form
      class="language-picker__form"
      data-testid="language-selector"
    >
      <div
        class="language-picker__group"
      >
        <label
          class="visually-hidden language-picker__label form-label"
          data-testid="lang-select-label"
          for="language-picker-select"
        >
          language.selector_label
        </label>
        <select
          aria-label="language.selector_label"
          class="language-picker__select form-select"
          data-testid="language-picker-select"
          id="language-picker-select"
          name="language-picker-select"
        >
          <option
            data-testid="language-picker__option_es-ES"
            value="es-ES"
          >
            language.es-ES
          </option>
          <option
            data-testid="language-picker__option_en-GB"
            value="en-GB"
          >
            language.en-GB
          </option>
        </select>
        <span
          class="visually-hidden"
          id="language-selector-desc"
        >
          language.selector_hint
        </span>
        <div
          aria-atomic="true"
          aria-live="polite"
          class="visually-hidden"
        >
          language.changed
        </div>
        <button
          aria-controls="language-picker-dropdown"
          aria-expanded="false"
          aria-label="English, Select your language"
          class="visually-hidden language-picker__button btn btn-primary"
          type="submit"
        >
          <span
            aria-hidden="true"
            class="language-picker__flag language-picker__flag--english"
          />
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
